File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/game/WordQuizEngine.kt
package org.nihongo.mochi.domain.game

import org.nihongo.mochi.domain.models.GameStatus
import org.nihongo.mochi.domain.models.Word

class WordQuizEngine {
    var isGameInitialized = false
    var allWords = mutableListOf<Word>()
    var currentWordSet = mutableListOf<Word>()
    var revisionList = mutableListOf<Word>()
    val wordStatus = mutableMapOf<Word, GameStatus>()
    var wordListPosition = 0
    var currentWord: Word? = null
    var currentAnswers = listOf<String>()
}

File: flutter_app/lib/services/word_quiz_engine.dart
import '../models/quiz_models.dart';

class WordQuizEngine {


  bool isGameInitialized = false;
  List<WordQuizItem> allWords = [];
  List<WordQuizItem> currentWordSet = [];
  List<WordQuizItem> revisionList = [];
  final Map<String, GameStatus> wordStatus = {};
  int wordListPosition = 0;
  WordQuizItem? currentWord;
  List<String> currentAnswers = [];

  void reset() {
    isGameInitialized = false;
    allWords = [];
    currentWordSet = [];
    revisionList = [];
    wordStatus.clear();
    wordListPosition = 0;
    currentWord = null;
    currentAnswers = [];
  }

  bool startNewSet() {
    revisionList.clear();
    wordStatus.clear();

    if (wordListPosition >= allWords.length) {
      return false;
    }

    final nextSet = allWords.skip(wordListPosition).take(10).toList();
    wordListPosition += nextSet.length;

    currentWordSet = nextSet;
    revisionList = List.from(nextSet);
    for (var word in nextSet) {
      wordStatus[word.text] = GameStatus.notAnswered;
    }
    return true;
  }
}
